# **Step-Wise Regression**

In [1]:
import numpy as np
import pandas as pd

In [2]:
from sklearn.datasets import load_diabetes
data = load_diabetes()

In [3]:
data.keys()

dict_keys(['data', 'target', 'frame', 'DESCR', 'feature_names', 'data_filename', 'target_filename', 'data_module'])

In [4]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from sklearn.model_selection import train_test_split

In [5]:
r2_scores = {}

In [6]:
for feature in range (len(data.feature_names)):
    X = data.data[:,feature]
    y = data.target
    X_train , X_test , y_train , y_test = train_test_split(X ,y , test_size=0.25 , random_state=42)
    X_train = np.array(X_train).reshape(-1, 1)
    X_test = np.array(X_test).reshape(-1, 1)
    reg = LinearRegression()
    reg.fit(X_train , y_train)
    y_pred = reg.predict(X_test)
    r2 = r2_score(y_test , y_pred)
    r2_scores.update({f"{feature}": f"{r2}"})

In [7]:
r2_scores

{'0': '0.008637038226036675',
 '1': '-0.01029825562499287',
 '2': '0.3172099449537781',
 '3': '0.18471383018307064',
 '4': '0.06626553266368607',
 '5': '0.04163111831069033',
 '6': '0.1377356557444941',
 '7': '0.19257590747667053',
 '8': '0.4215743297548834',
 '9': '0.16593419294983636'}

In [8]:
sorted_dict = sorted(r2_scores.items(), key=lambda x:x[1] , reverse=True)
for feature in sorted_dict:
    print(feature)


('8', '0.4215743297548834')
('2', '0.3172099449537781')
('7', '0.19257590747667053')
('3', '0.18471383018307064')
('9', '0.16593419294983636')
('6', '0.1377356557444941')
('4', '0.06626553266368607')
('5', '0.04163111831069033')
('0', '0.008637038226036675')
('1', '-0.01029825562499287')


In [9]:
feature_idx = [8,2,7,3,9,6,4,5,0,1]

In [10]:
data = load_diabetes(as_frame=True)

# **Forward-Selection**

In [11]:
selected_features = data.data.iloc[:, [8,2]]
X = selected_features.values
y = data.target

In [12]:
X_train , X_test , y_train , y_test = train_test_split(X ,y , test_size=0.25 , random_state=42)
reg = LinearRegression()
reg.fit(X_train , y_train)
y_pred = reg.predict(X_test)
r2 = r2_score(y_test , y_pred)
print(f"R2 score after combining : {r2}")

R2 score after combining : 0.5112943322988956


In [13]:
selected_features = data.data.iloc[:, [8,2,7]]
X = selected_features.values
y = data.target
X_train , X_test , y_train , y_test = train_test_split(X ,y , test_size=0.25 , random_state=42)
reg = LinearRegression()
reg.fit(X_train , y_train)
y_pred = reg.predict(X_test)
r2 = r2_score(y_test , y_pred)
print(f"R2 score after combining : {r2}")

R2 score after combining : 0.5107349369399403


In [14]:
selected_features = data.data.iloc[:, [8,2,7,3]]
X = selected_features.values
y = data.target
X_train , X_test , y_train , y_test = train_test_split(X ,y , test_size=0.25 , random_state=42)
reg = LinearRegression()
reg.fit(X_train , y_train)
y_pred = reg.predict(X_test)
r2 = r2_score(y_test , y_pred)
print(f"R2 score after combining : {r2}")

R2 score after combining : 0.49776129259653035


**We find that the r2 square value is not increasing so we stop here. the best performing subset has an r2 square value of 0.5112943322988956**

# **Backward-Elimination**

In [15]:
X = data.data
y = data.target
X_train , X_test , y_train , y_test = train_test_split(X ,y , test_size=0.25 , random_state=42)
reg = LinearRegression()
reg.fit(X_train , y_train)
y_pred = reg.predict(X_test)
r2 = r2_score(y_test , y_pred)
print(f"R2 score after combining : {r2}")

R2 score after combining : 0.4849058889476755


In [16]:
selected_features = data.data.iloc[:, [8,2,7,3,9,6,4,5,0]]
X = selected_features.values
y = data.target
X_train , X_test , y_train , y_test = train_test_split(X ,y , test_size=0.25 , random_state=42)
reg = LinearRegression()
reg.fit(X_train , y_train)
y_pred = reg.predict(X_test)
r2 = r2_score(y_test , y_pred)
print(f"R2 score after combining : {r2}")

R2 score after combining : 0.4705871168973681


In [17]:
selected_features = data.data.iloc[:, [8,2,7,3,9,6,4,5]]
X = selected_features.values
y = data.target
X_train , X_test , y_train , y_test = train_test_split(X ,y , test_size=0.25 , random_state=42)
reg = LinearRegression()
reg.fit(X_train , y_train)
y_pred = reg.predict(X_test)
r2 = r2_score(y_test , y_pred)
print(f"R2 score after combining : {r2}")

R2 score after combining : 0.47248980041817334


In [18]:
selected_features = data.data.iloc[:, [8,2,7,3,9,6,4]]
X = selected_features.values
y = data.target
X_train , X_test , y_train , y_test = train_test_split(X ,y , test_size=0.25 , random_state=42)
reg = LinearRegression()
reg.fit(X_train , y_train)
y_pred = reg.predict(X_test)
r2 = r2_score(y_test , y_pred)
print(f"R2 score after combining : {r2}")

R2 score after combining : 0.47492399586539447


In [19]:
selected_features = data.data.iloc[:, [8,2,7,3,9,6]]
X = selected_features.values
y = data.target
X_train , X_test , y_train , y_test = train_test_split(X ,y , test_size=0.25 , random_state=42)
reg = LinearRegression()
reg.fit(X_train , y_train)
y_pred = reg.predict(X_test)
r2 = r2_score(y_test , y_pred)
print(f"R2 score after combining : {r2}")

R2 score after combining : 0.4782057877081908


In [20]:
selected_features = data.data.iloc[:, [8,2,7,3,9]]
X = selected_features.values
y = data.target
X_train , X_test , y_train , y_test = train_test_split(X ,y , test_size=0.25 , random_state=42)
reg = LinearRegression()
reg.fit(X_train , y_train)
y_pred = reg.predict(X_test)
r2 = r2_score(y_test , y_pred)
print(f"R2 score after combining : {r2}")

R2 score after combining : 0.4984370023711241


In [21]:
selected_features = data.data.iloc[:, [8,2,7,3]]
X = selected_features.values
y = data.target
X_train , X_test , y_train , y_test = train_test_split(X ,y , test_size=0.25 , random_state=42)
reg = LinearRegression()
reg.fit(X_train , y_train)
y_pred = reg.predict(X_test)
r2 = r2_score(y_test , y_pred)
print(f"R2 score after combining : {r2}")

R2 score after combining : 0.49776129259653035


In [22]:
selected_features = data.data.iloc[:, [8,2,7]]
X = selected_features.values
y = data.target
X_train , X_test , y_train , y_test = train_test_split(X ,y , test_size=0.25 , random_state=42)
reg = LinearRegression()
reg.fit(X_train , y_train)
y_pred = reg.predict(X_test)
r2 = r2_score(y_test , y_pred)
print(f"R2 score after combining : {r2}")

R2 score after combining : 0.5107349369399403


In [23]:
selected_features = data.data.iloc[:, [8,2]]
X = selected_features.values
y = data.target
X_train , X_test , y_train , y_test = train_test_split(X ,y , test_size=0.25 , random_state=42)
reg = LinearRegression()
reg.fit(X_train , y_train)
y_pred = reg.predict(X_test)
r2 = r2_score(y_test , y_pred)
print(f"R2 score after combining : {r2}")

R2 score after combining : 0.5112943322988956


**We arrive to the same subset of features as forward-selection**

**Step-wise regression is a combination of both these approaches , both these approaches don't do feature selection only sequentially , but randomly also**